File: app/src/main/java/com/artamonov/millionplanets/NewGameActivity.java
package com.artamonov.millionplanets;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import androidx.appcompat.app.AppCompatActivity;
import com.artamonov.millionplanets.model.User;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Transaction;
import java.util.HashMap;
import java.util.Map;

public class NewGameActivity extends AppCompatActivity {
    FirebaseAuth firebaseAuth;
    EditText etNickname;
    FirebaseFirestore firebaseFirestore;
    String username;
    private DocumentReference documentReferenceInventory;
    private DocumentReference documentReferenceObjects;
    private FirebaseUser firebaseUser;
    private User user;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.create_player_activity);
        etNickname = findViewById(R.id.register_username);
        firebaseFirestore = FirebaseFirestore.getInstance();
    }

    public void onGoToSpace(View view) {
        if (!nicknameIsValid()) {
            Snackbar.make(
                            findViewById(android.R.id.content),
                            "Your nickname must be at least 3 characters",
                            Snackbar.LENGTH_SHORT)
                    .show();
            return;
        }
        firebaseAuth = FirebaseAuth.getInstance();
        username = etNickname.getText().toString();
        firebaseUser = firebaseAuth.getCurrentUser();
        documentReferenceInventory = firebaseFirestore.collection("Inventory").document(username);
        documentReferenceObjects = firebaseFirestore.collection("Objects").document(username);
        user = new User();
        user.setX(5);
        user.setY(6);
        user.setCargo(10);
        user.setHp(50);
        user.setShip("Fighter");
        user.setMoney(1000);
        user.setScanner_capacity(15);
        user.setShield(100);
        user.setJump(10);
        user.setFuel(20);
        user.setNickname(username);
        user.setEmail(firebaseUser.getEmail());
        user.setType("user");
        user.setSumXY(11);
        // Log.i("myTags", "nickname: " + firebaseUser.getDisplayName());
        /*  firebaseFirestore.collection("Objects").document(username)
                        .set(user).addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        UserProfileChangeRequest profileUpdates = new UserProfileChangeRequest.Builder()
                                .setDisplayName(username)
                                .build();
                        firebaseUser.updateProfile(profileUpdates).addOnSuccessListener(new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                startActivity(new Intent(getApplicationContext(), MainOptionsActivity.class));
                            }
                        });

                    }
                });
        */

        firebaseFirestore
                .runTransaction(
                        new Transaction.Function<Void>() {

                            @Override
                            public Void apply(Transaction transaction) {
                                transaction.set(documentReferenceObjects, user);
                                UserProfileChangeRequest profileUpdates =
                                        new UserProfileChangeRequest.Builder()
                                                .setDisplayName(username)
                                                .build();
                                firebaseUser.updateProfile(profileUpdates);
                                Map<String, Object> ironData = new HashMap<>();
                                ironData.put("Iron", 0);
                                ironData.put("Mercaster", 0);
                                ironData.put("Leabia", 0);
                                ironData.put("Cracaphill", 0);
                                transaction.set(documentReferenceInventory, ironData);
                                return null;
                            }
                        })
                .addOnSuccessListener(
                        new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                startActivity(
                                        new Intent(
                                                getApplicationContext(),
                                                MainOptionsActivity.class));
                            }
                        });

        /* private void setGeoData() {
            ObjectModel objectModel = new ObjectModel();
            objectModel.setX(user.getX());
            objectModel.setY(user.getY());
            objectModel.setSumXY(user.getSumXY());
            objectModel.setType("user");

            firebaseFirestore.collection("GeoData").document(username)
                    .set(objectModel).addOnSuccessListener(new OnSuccessListener<Void>() {
                @Override
                public void onSuccess(Void aVoid) {
                    startActivity(new Intent(getApplicationContext(), MainOptionsActivity.class));
                }
            });
        }*/

    }

    private boolean nicknameIsValid() {
        return etNickname.getText().length() > 2;
    }
}

File: app/src/main/java/com/artamonov/millionplanets/NewGameActivity.kt
package com.artamonov.millionplanets

import android.content.Intent
import android.os.Bundle
import android.os.PersistableBundle

import com.artamonov.millionplanets.base.BaseActivity
import com.artamonov.millionplanets.databinding.ActivityNewGameBinding
import com.artamonov.millionplanets.model.SpaceshipType
import com.artamonov.millionplanets.model.User
import com.artamonov.millionplanets.utils.showSnackbarError
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.firestore.DocumentReference
import java.util.HashMap

class NewGameActivity : BaseActivity() {
    private lateinit var username: String
    private var documentReferenceInventory: DocumentReference? = null
    private var documentReferenceObjects: DocumentReference? = null
    private var user: User? = null

    lateinit var binding: ActivityNewGameBinding

    override fun onCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        super.onCreate(savedInstanceState, persistentState)
        binding = ActivityNewGameBinding.inflate(layoutInflater)
        setContentView(binding.root)
        binding.goToSpace.setOnClickListener {
            onGoToSpace()
        }
    }

    private fun onGoToSpace() {
        if (!nicknameIsValid()) {
            showSnackbarError(getString(R.string.registration_nickname_must_at_least_3_characters_long))
            return
        }

        firebaseAuth = FirebaseAuth.getInstance()
        username = binding.registerUsername.text.toString()
        documentReferenceInventory = firebaseFirestore.collection("Inventory").document(username)
        documentReferenceObjects = firebaseFirestore.collection("Objects").document(username)
        user = User(x = 5, y = 6, cargoCapacity = 10, hp = 50, ship = SpaceshipType.FIGHTER,
                money = 1000, scanner_capacity = 15, shield = 100, jump = 10, fuel = 20, nickname =
        username, email = firebaseUser.email, type = "user", sumXY = 11)

        // Log.i("myTags", "nickname: " + firebaseUser.getDisplayName());
        /*  firebaseFirestore.collection("Objects").document(username)
                        .set(user).addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        UserProfileChangeRequest profileUpdates = new UserProfileChangeRequest.Builder()
                                .setDisplayName(username)
                                .build();
                        firebaseUser.updateProfile(profileUpdates).addOnSuccessListener(new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                startActivity(new Intent(getApplicationContext(), MainOptionsActivity.class));
                            }
                        });

                    }
                });
        */

        firebaseFirestore
                .runTransaction { transaction ->
                    transaction.set(documentReferenceObjects!!, user!!)
                    val profileUpdates = UserProfileChangeRequest.Builder()
                            .setDisplayName(username)
                            .build()
                    firebaseUser.updateProfile(profileUpdates)
                    val ironData = HashMap<String, Any>()
                    ironData["Iron"] = 0
                    ironData["Mercaster"] = 0
                    ironData["Leabia"] = 0
                    ironData["Cracaphill"] = 0
                    transaction.set(documentReferenceInventory!!, ironData)
                    null
                }
                .addOnSuccessListener {
                    startActivity(
                            Intent(this,
                                    MainOptionsActivity::class.java))
                }
    }

    private fun nicknameIsValid(): Boolean {
        return binding.registerUsername.text.length > 2
    }
}
